toolchain/base/value_store_chunk.h

```diff
@@ -0,0 +1,175 @@
+  // PlatformChunkCapacity(). This ensures that ids are contiguous values
+  // from 0, as if the values were all stored in a single array, and allows
+  // using the ids to index into other arrays.
+  static_assert((1 << LowBits) == PlatformChunkCapacity<IdT>());
+  // Simple check to make sure nothing went wildly wrong with the
+  // PlatformChunkCapacity, and we have some room for a chunk index, and
+  // that shifting by the number of bits won't be UB in an int32_t.
+  static_assert(LowBits < 30);
+
+  // We can only use 31 bits in total because the sign bit is used for making
+  // negative ids, which are not found in the ValueStore.
+  constexpr auto HighBits = 31 - LowBits;
+
+  // The index of the chunk is the high bits.
+  auto chunk = id.index >> LowBits;
+  // The index into the chunk is the low bits.
+  auto pos = id.index & ((1 << LowBits) - 1);
+
+  // This routine is especially hot and the check here relatively expensive for
+  // the value provided, so only do this in non-optimized builds to make
+  // tracking down issues easier.
+  CARBON_DCHECK(chunk < (1 << HighBits), "Id overflow (high bits)");
+  CARBON_DCHECK(pos < (1 << LowBits), "Id overflow (low bits)");
+
+  return {chunk, pos};
+}
+
+// A chunk of `ValueType`s which has a fixed capacity, but variable size. Tracks
+// the size internally for verifying bounds.
+template <typename IdT, class ValueType>
+struct ValueStoreChunk {
+ public:
+  static constexpr auto Capacity = Internal::PlatformChunkCapacity<IdT>();
+  static constexpr auto CapacityBytes = Capacity * sizeof(ValueType);
+
+  explicit ValueStoreChunk()
+      : buf_(reinterpret_cast<ValueType*>(
+            llvm::allocate_buffer(CapacityBytes, alignof(ValueType)))) {}
+
+  // Moving leaves nullptr behind in the moved-from object so that the
+  // destructor is a no-op (preventing double free).
+  ValueStoreChunk(ValueStoreChunk&& rhs) noexcept
+      : buf_(std::exchange(rhs.buf_, nullptr)), num_(rhs.num_) {}
+
+  auto operator=(ValueStoreChunk&& rhs) noexcept -> ValueStoreChunk& {
+    buf_ = std::exchange(rhs.buf_, nullptr);
+    num_ = rhs.num_;
+    return *this;
+  }
+
+  ~ValueStoreChunk() {
+    if (buf_) {
+      if constexpr (!std::is_trivially_destructible_v<ValueType>) {
+        std::destroy_n(buf_, num_);
+      }
+      llvm::deallocate_buffer(buf_, CapacityBytes, alignof(ValueType));
+    }
+  }
+
+  auto at(int32_t i) -> ValueType& {
+    CARBON_CHECK(i < num_, "{0}", i);
+    return buf_[i];
+  }
+  auto at(int32_t i) const -> const ValueType& {
+    CARBON_CHECK(i < num_, "{0}", i);
+    return buf_[i];
+  }
+
+  auto push(ValueType&& value) -> void {
+    CARBON_CHECK(num_ < Capacity);
+    std::construct_at(buf_ + num_, std::move(value));
+    ++num_;
+  }
+
+  auto size() const -> int32_t { return num_; }
+
+ private:
+  // Verify using an `int32_t` for `num_` is sound.
+  static_assert(Capacity <= std::numeric_limits<int32_t>::max());
+
+  ValueType* buf_;
+  int32_t num_ = 0;
```

Comment:
https://github.com/carbon-language/carbon-lang/pull/5576#discussion_r2114998169 brought up removing this size -- should we discuss that further in a meeting? I'm fine with this for now, but I think the duplication is something we should remove.

Reply:
Let's discuss. I think it makes our code much more prone to buffer overflow mistakes.